File: src/main/java/me/laus/tallinn/transport/client/external/RealTimeClient.java
package me.laus.tallinn.transport.client.external;

import me.laus.tallinn.transport.model.realtime.Vehicle;
import me.laus.tallinn.transport.model.request.HttpMethod;

import java.net.http.HttpRequest;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class RealTimeClient extends TallinnExternalApiClient {
    private static final String GPS_ENDPOINT = "/gps.txt";

    public RealTimeClient() {
        super(GPS_ENDPOINT);
    }

    public List<Vehicle> fetch() {
        HttpRequest request = buildRequest(HttpMethod.GET);
        String[] response = sendRequest(request);
        try {
            return Arrays
                    .stream(response)
                    .map(position -> position.split(","))
                    .map(Vehicle.Factory::fromList)
                    .toList();
        } catch (IllegalArgumentException e) {
            return Collections.emptyList();
        }
    }
}

File: src/main/java/me/laus/tallinn/transport/client/external/RealTimeClient.kt
package me.laus.tallinn.transport.client.external

import me.laus.tallinn.transport.model.realtime.Vehicle
import me.laus.tallinn.transport.model.request.HttpMethod
import java.util.*

class RealTimeClient : TallinnExternalApiClient(GPS_ENDPOINT) {
    fun fetch(): List<Vehicle> {
        val request = buildRequest(HttpMethod.GET)
        val response = sendRequest(request)
        return try {
            Arrays
                .stream(response)
                .map { position: String? -> position!!.split(",").toTypedArray() }
                .map { parts: Array<String>? -> parts?.let { Vehicle.Factory.fromList(it) } }
                .toList()
        } catch (e: IllegalArgumentException) {
            emptyList()
        }
    }

    companion object {
        private const val GPS_ENDPOINT = "/gps.txt"
    }
}